WIP!! Import integrations initial setup, first pass

kahboom opened the pull request:
Not that you would, but please don't merge this in yet. Quite a bit of outstanding work; leaving open for collaboration and just to show what I have so far.

Some issues:
- Resolve merge conflicts.
- Type errors resulting from incorrect, well, typings and issues naming things. And also it's 3:30am.
- Still getting used to effects, redux containers, and deciding whether or not to have a store for certain things, so these areas need to be worked on a bit.
- Naming! Importing and uploading seemed sufficient, but I wasn't sure whether or not to include something as detailed as ImportIntegrationImport or ImportIntegrationUpload.
- Templating! I'm sure the app won't build properly because of the aforementioned reasons.
- Some confusion about where there should be overlap between import and exporting, and well as multi versus single import.
- No proper tests.
- Among other things.

I've taken some parts of the API Custom Connector and adjusted them for this particular feature, and that is still undergoing work. Reducer and actions have been written accordingly, but likely contain issues. Tried to check the Swagger JSON on staging with not much luck for data structure, and I don't think we have design proposals anymore, so I guess I'll see about that tomorrow.

@deeleman - Please take a look, but also don't kill me. This has a long way to go.

deeleman commented on app/ui/src/app/import-export/import-export.component.scss:
```diff
@@ -0,0 +1,24 @@
+  overflow: visible;
+}
+
+ul {
```
You do not need this `ul` styling anymore. Jas add the `syn-tabs` class to the navbar and some additional classnames to the listitems and you're all sel. 

Double check the example in the Monitor metrics for reference :)

deeleman commented on app/ui/src/app/import-export/import-export.module.ts:
```diff
@@ -0,0 +1,41 @@
+      }
+    ]
+  }
+];
```
It' usually a good practice to promote routing to its own routing module, and import it inot the gneeral domain module. Check the Metrics module for an example and download the John Papa's Angular Snippets for autogenerating that code in your IDE

deeleman commented on app/ui/src/app/import-export/import/import.actions.ts:
```diff
@@ -0,0 +1,117 @@
+
+  /**
+   * Not actually sure that we would need fetching here.
+   */
```
It's always better to start with what you really need and then go adding stuff as you need it. Otherwise reducers can become massive and kinda confusing ;)

So, if you see there're actions you're not fully confident that you'll want to use, remove those

deeleman commented on app/ui/src/app/import-export/import/import.actions.ts:
```diff
@@ -0,0 +1,117 @@
+  static uploadCancel(): IntegrationUploadCancel {
+    return new IntegrationUploadCancel();
+  }
+}
```
The idea for creating static methods in a Actions class to dispatch actions more easily eventually proved to generate too much boilerplate. Perhaps is better to use Action classes as is. Again, you can see the leaner approach I took for the metrics domain asa reference.

deeleman commented on app/ui/src/app/import-export/import/import.component.scss:
```diff
@@ -0,0 +1,25 @@
+  flex: 0 0 auto;
+}
+
+syndesis-loading {
```
Isn't this piece of SASS used elsewhere? If so, perhaps it would be a good idea to promote it to `src/scss/components/_loading.scss`

deeleman commented on app/ui/src/app/import-export/import/import.component.ts:
```diff
@@ -0,0 +1,68 @@
+import {
+  FileUploader,
+  FileItem
+} from 'ng2-file-upload';
```
No need for a breakline, it's just two classes ;P

deeleman commented on app/ui/src/app/import-export/import/import.component.ts:
```diff
@@ -0,0 +1,68 @@
+  styleUrls: ['./import-integration.component.scss']
+})
+export class ImportIntegrationComponent implements OnInit {
+  public uploader: FileUploader;
```
In TypeScript all members are public by default, so this accessor is redundant. not wrong, but redundant.

deeleman commented on app/ui/src/app/import-export/import/import.component.ts:
```diff
@@ -0,0 +1,68 @@
+  public uploader: FileUploader;
+
+  loading: Observable<boolean>;
+  integrations: Observable<Integrations>;
```
I'd suggest to conform to the `loading$` or `integrations$` naming convention.

deeleman commented on app/ui/src/app/import-export/import/index.ts:
```diff
@@ -0,0 +1,5 @@
+export * from './import.actions';
+export * from './import.reducer';
+export * from './import.models';
+export * from './import.module';
```
models, actions and reducer are most likely unnecessary outside the scope of this module, so there is no need to expose those.

deeleman commented on app/ui/src/app/import-export/import/import.reducer.ts:
```diff
@@ -0,0 +1,142 @@
+  IntegrationImportComplete,
+  IntegrationImportFail,
+} from './import.actions';
+
```
`import * as ImportActions from './import.actions';`is simpler and more scalable, so you won't have to import each new action yo create in the future.

gashcrumb commented:
@kahboom run `yarn build:ci` locally to catch any AOT issues, speeds up fixing them a bit.

chirino commented on app/ui/src/app/integration/import-export/integration-import-export.component.html:
```diff
@@ -0,0 +1,7 @@
+<ul class="row nav nav-tabs nav-tabs-pf toolbar-pf">
+  <li routerLinkActive="active">
+    <a [routerLink]="['api-connector']">Import Integration</a>
```
'api-connector' seems like an odd link for import integration.

kahboom replied:
This is copy-pasta.

chirino commented on app/ui/src/app/integration/import-export/import/integration-import.service.ts:
```diff
@@ -0,0 +1,51 @@
+export class IntegrationImportService {
+  constructor(private apiHttpService: ApiHttpService) { }
+
+  getImport(id: string): Observable<IntegrationImportData> {
```
Don't think we need ti s class, we can't upload then review for imports.  Once the file is summitted to the importer, your done.  It's imported.  You just need to display the import results.

kahboom replied:
I am removing the `fetch`-related Redux actions, hence this being a WIP.

chirino commented on app/ui/src/app/integration/import-export/import/integration-import.component.ts:
```diff
@@ -0,0 +1,65 @@
+      response: string,
+      status: number
+    ) => {
+      if (status === 204) {
```
Don't you have to take this result and put it in the result page of the wizard instead of poping up notifications?

deeleman commented on app/ui/src/app/integration/import-export/import/index.ts:
```diff
@@ -0,0 +1,5 @@
+export * from './integration-import.module';
+export * from './integration-import.actions';
+export * from './integration-import.models';
+export * from './integration-import.reducer';
```
You're already importing and bootstrapping the Reducer at `integration-import.module`, hence there's no need to export it from here. all the way around, this might end up in a circular reference. Same applies to `integration-import.component` above.

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.actions.ts:
```diff
@@ -0,0 +1,41 @@
+ * Once the integration has been uploaded, it is available for review.
+ * Once the user has reviewed the payload, they can then press 'Done' to
+ * proceed to the Edit Integration view, which will be populated.
+ */
```
You're describing a UI interaction here. Actions are meant to be UI-agnostic. They just deal with data/state operations.

kahboom replied:
It was for reference so that developers know what this action is typically used for, but I do see what you mean. I'll remove it.

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.component.html:
```diff
@@ -0,0 +1,21 @@
+  Select the integration file to import:
+  <form class="syn-form syn-form--integration-upload"
+        (ngSubmit)="onSubmit(integrationImportForm, uploadFileOption.checked)"
+        #swaggerForm="ngForm"
```
Is the integration import file coming as a swagger file?

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.component.scss:
```diff
@@ -0,0 +1,25 @@
+
+.max-height {
+  height:100%
+}
```
This chunk of SASS basically duplicates to the point what we already had at `app/integration/list-page/list-page.component.scss`. This kind of code duplication must be avoided by all means, since it only increases the overall footprint of the application bundle and prevents CSS reusability and maintainability. You have several workarounds for this:

- Move this chunk of code to a shared location and import it from both components (remember that the `styleUrls` property expects an array with not one but many references).
- Turn this piece of code into a SASS partial and import it from each component's own stylesheet, should you feel that each component will eventually feature its own CSS nuances.
- Move parts of the code to SASS `%` extension classnames and use them internally within these selectors.
- Use a global mixin and import it directly.

In addition to the approaches above, you should strive to stick to the app variables we defined at `/scss/base/_variables.scss` in order to keep consistency in the the UI's look and feel. So whenever you're dealing with component CSS variables, you should strive to avoid hardcoded values as much as you can.

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.models.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+export type IntegrationImportsEditState = Array<IntegrationImportEditState>;
+
```
What is the point of defining two different state interfaces for the same schema? And two types for an interface array with different names?

deeleman replied:
If you want to define an interface that models the import operation (regardless it models an upload or edit operation) just do it once! And strive for interface reusability whenever you can (as in this place).

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.effects.ts:
```diff
@@ -0,0 +1,34 @@
+export class IntegrationImportEffects {
+  @Effect()
+  importIntegration$: Observable<Action> = this.actions$
+    .ofType<IntegrationUpload>(IntegrationImportActions.IMPORT)
```
You can spare importing `IntegrationUpload ` and use `IntegrationImportActions .IntegrationUpload ` here instead. It might be more verbose but will allow you to not to keep track and import each and every action created in the future.

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.effects.ts:
```diff
@@ -0,0 +1,34 @@
+  IntegrationUpload,
+  IntegrationUploadComplete,
+  IntegrationImport,
+  IntegrationImportComplete
```
Half of these tokens are not in use

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.component.ts:
```diff
@@ -0,0 +1,35 @@
+  ngOnInit() {
+    /**
+     * Payload for IntegrationImportUpload action
+     */
```
Why do we need a comment for this? Isn't the variable names or the method naming self-descriptive enough?

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.component.ts:
```diff
@@ -0,0 +1,35 @@
+
+
+    // We dispatch the desired action with the given payload
+    this.store.dispatch(new IntegrationImportUpload(payload));
```
Do you believe you're gonna be needing to dispatch the `IntegrationImportUpload` with different payloads in the future. Otherwise there is no point to dispatch an action with an object payload that reflects and empty initial state. You already have `initialState` at the reducer for this, and you can even initialize empty instances of state within the reducer handlers if needed.

Here we're only duplicating code and action payloads should be as minimal as possible or even non-existing whereas possible.

deeleman commented on app/ui/src/app/integration/import-export/import/integration-import.service.ts:
```diff
@@ -0,0 +1,45 @@
+      importRequest,
+      importRequest.iconFile,
+      importRequest.specificationFile
+    ];
```
Beware with copy+pasted code, since that's the easiest way to introduce technical debt into a project: `iconfile` or `specificationFile` do not exist in the `IntegrationImportRequest` interface.

kahboom replied:
I was trying to use the ApiConnector stuff as an example for following the Redux architecture, but ended up copying (and subsequently removing) a lot of the code. I think it would have been best to apply what I'm actively learning outside of a feature requirement.

deeleman commented on app/ui/src/app/integration/import-export/integration-import-export.component.scss:
```diff
@@ -0,0 +1,24 @@
+      cursor: pointer;
+    }
+  }
+}
```
All the `<ul>` styling here has been thoroughly duplicated all over the project and in order to prevent more duplication we now have a global classname named `.syn-tabs` that takes care of that for the entire project. FYI it is available at `/scss/components/_tabbed-nav.scss` and you do not need to import `syndesis-sass` to use it in your HTML because it is already part of the CSS classnames bndle.

Please avoid SASS duplication.

kahboom commented:
Squashed a bunch of commits, fixed the lint errors and failing build, made changes according to feedback. I think I set up the Redux stuff as best as I could--only concern is the integration routing seems wonky, so I'm sorting that out. Adding actions, effects, etc. as necessary for the editor (step 2) will come later on.

A separate PR it will be heavily focused on API wiring and component/template (actually implementing the feature) using this architecture will be opened once this one looks okay.

kahboom commented:
This is ready for review again. Lint and tests pass, AOT build is fine, fixed the issue with the routing. This is for the architectural setup, I'm already working on opening the feature implementation PR. For more information, please see the primary issue.

kahboom commented:
There is a `<h1></h1>` and `<h2></h2>` tag visible on the page now from when I was testing the templating and routes, but since it's not a central page on the app I'd prefer to leave it there for this PR--it will be removed shortly with the next PR. See screenshot below:

<img width="1680" alt="screenshot 2018-02-14 13 53 51" src="https://user-images.githubusercontent.com/3844502/36222212-a656b1f4-118e-11e8-947d-c790e59f952d.png">

kahboom commented:
It looks like the REST build failed for some reason; there is notice on circleCI for partially degraded service that suggests potential intermittent failures. I don't have permission to trigger a build manually, but will try something.
